File: javascript/easy/ransomNote.js
/**
 * @param {string} ransomNote
 * @param {string} magazine
 * @return {boolean}
 */
var canConstruct = function(ransomNote, magazine) {
    const map = new Map();
    for(let c of ransomNote) {
        let count = map.get(c); 
        map.set(c, ((count === undefined) ? 1 : count+1));
    }
    
    for(let c of magazine) {
        let count = map.get(c);
        if(count !== undefined) {
            if(count === 1) {
                map.delete(c);
            } else {
                map.set(c, count-1);
            }
        }
    }
    
    return map.size === 0;
};

File: ruby/easy/ransom_note.rb
# @param {String} ransom_note
# @param {String} magazine
# @return {Boolean}
def can_construct(ransom_note, magazine)
    map = {}
    ransom_note.each_char do |c|
        map[c] = map[c] ? (map[c] + 1) : 1
    end
    magazine.each_char do |c|
        map[c] = map[c] - 1 if map[c]
        map.delete(c) if map[c] == 0
    end
    return map.length == 0
end
